go/libraries/doltcore/doltdb/doltdb.go

```diff
@@ -1392,3 +1392,146 @@ func (ddb *DoltDB) GetBranchesByRootHash(ctx context.Context, rootHash hash.Hash
+	if err != nil {
+		return err
+	}
+
+	stashListAddr, err := stashList.AddStash(ctx, vrw, stashAddr)
+	if err != nil {
+		return err
+	}
+
+	stashesDS, err = ddb.db.UpdateStashList(ctx, stashesDS, stashListAddr)
+	return err
+}
+
+// RemoveStashAtIdx takes and index of a stash to remove from the stash list map.
+// It removes a Stash message from stash list Dataset, which cannot be performed
+// by database Delete function. This function removes a single stash only and stash
+// list dataset does not get removed if there are no entries left.
+func (ddb *DoltDB) RemoveStashAtIdx(ctx context.Context, idx int) error {
+	stashesDS, err := ddb.db.GetDataset(ctx, ref.NewStashRef().String())
+	if err != nil {
+		return err
+	}
+
+	if !stashesDS.HasHead() {
+		return errors.New("No stash entries found.")
+	}
+
+	vrw := ddb.ValueReadWriter()
+	stashList, err := datas.LoadStashList(ctx, ddb.Format(), ddb.NodeStore(), vrw, stashesDS)
+	if err != nil {
+		return err
+	}
+
+	stashListAddr, err := stashList.RemoveStashAtIdx(ctx, vrw, idx)
+	if err != nil {
+		return err
+	}
+
+	stashesDS, err = ddb.db.UpdateStashList(ctx, stashesDS, stashListAddr)
+	return err
+}
+
+// RemoveAllStashes removes all Stash entries from the stash list Dataset.
+// TODO: could we just delete the whole Dataset in this case?
+func (ddb *DoltDB) RemoveAllStashes(ctx context.Context) error {
+	stashesDS, err := ddb.db.GetDataset(ctx, ref.NewStashRef().String())
+	if err != nil {
+		return err
+	}
+
+	if !stashesDS.HasHead() {
+		return nil
+	}
+
+	vrw := ddb.ValueReadWriter()
+	stashList, err := datas.LoadStashList(ctx, ddb.Format(), ddb.NodeStore(), vrw, stashesDS)
+	if err != nil {
+		return err
+	}
+
+	stashListAddr, err := stashList.ClearAllStashes(ctx, vrw)
+	if err != nil {
+		return err
+	}
+
+	stashesDS, err = ddb.db.UpdateStashList(ctx, stashesDS, stashListAddr)
+	return err
+}
+
+// GetStashes returns array of Stash objects containing all stash entries in the stash list Dataset.
+func (ddb *DoltDB) GetStashes(ctx context.Context) ([]*Stash, error) {
+	stashesDS, err := ddb.db.GetDataset(ctx, ref.NewStashRef().String())
+	if err != nil {
+		return nil, err
+	}
+
+	if !stashesDS.HasHead() {
+		return []*Stash{}, nil
+	}
+
+	return getStashList(ctx, stashesDS, ddb.vrw, ddb.NodeStore())
+}
+
+// GetStashHashAtIdx returns hash address only of the stash at given index.
+func (ddb *DoltDB) GetStashHashAtIdx(ctx context.Context, idx int) (hash.Hash, error) {
+	ds, err := ddb.db.GetDataset(ctx, ref.NewStashRef().String())
+	if err != nil {
+		return hash.Hash{}, err
+	}
+
+	if !ds.HasHead() {
+		return hash.Hash{}, errors.New("No stash entries found.")
+	}
+
+	return getStashHashAtIdx(ctx, ds, ddb.NodeStore(), idx)
+}
+
+// GetStashAtIdx returns root value of stash working set and head commit of the branch that the stash was made on
+// of the stash at given index.
+func (ddb *DoltDB) GetStashAtIdx(ctx context.Context, idx int) (*RootValue, *Commit, error) {
```

Comment:
(minor) Maybe `GetStashRootAtIdx`, since we aren't returning a `Stash` struct, but a `RootValue` (and `Commit`)?